File: python/mirage/commands/builtin/databricks_volume/rm.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.databricks_volume import DatabricksVolumeAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.databricks_volume.glob import resolve_glob
from mirage.core.databricks_volume.rm import rm_recursive
from mirage.core.databricks_volume.rmdir import rmdir
from mirage.core.databricks_volume.stat import stat
from mirage.core.databricks_volume.unlink import unlink
from mirage.io.types import ByteSource, IOResult
from mirage.types import FileType, PathSpec


@command("rm", resource="databricks_volume", spec=SPECS["rm"], write=True)
async def rm(
    accessor: DatabricksVolumeAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: bytes | None = None,
    r: bool = False,
    R: bool = False,
    f: bool = False,
    v: bool = False,
    d: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if not paths:
        raise ValueError("rm: missing operand")
    paths = await resolve_glob(accessor, paths, index)
    recursive = r or R
    verbose_parts: list[str] = []
    removed: dict[str, bytes] = {}
    for path in paths:
        try:
            file_stat = await stat(accessor, path, index)
        except FileNotFoundError:
            if f:
                continue
            raise
        if file_stat.type == FileType.DIRECTORY:
            if recursive:
                for relative in await rm_recursive(accessor, path, index):
                    removed[relative] = b""
            elif d:
                await rmdir(accessor, path, index)
                removed[path.strip_prefix] = b""
            else:
                raise IsADirectoryError(
                    f"rm: cannot remove '{path.original}': Is a directory")
        else:
            await unlink(accessor, path, index)
            removed[path.strip_prefix] = b""
        if v:
            verbose_parts.append(f"removed '{path.original}'")
    output = "\n".join(verbose_parts).encode() if v else None
    return output, IOResult(writes=removed)

File: typescript/packages/core/src/commands/builtin/databricks_volume/rm.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { DatabricksVolumeAccessor } from '../../../accessor/databricks_volume.ts'
import { resolveGlob } from '../../../core/databricks_volume/glob.ts'
import { readdir as dbxReaddir } from '../../../core/databricks_volume/readdir.ts'
import { rmRecursive as dbxRmR } from '../../../core/databricks_volume/rm.ts'
import { rmdir as dbxRmdir } from '../../../core/databricks_volume/rmdir.ts'
import { stat as dbxStat } from '../../../core/databricks_volume/stat.ts'
import { unlink as dbxUnlink } from '../../../core/databricks_volume/unlink.ts'
import { IOResult, type ByteSource } from '../../../io/types.ts'
import { FileType, type PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { formatRecords } from '../utils/output.ts'

const ENC = new TextEncoder()

interface RmOpts {
  recursive: boolean
  force: boolean
  removeDir: boolean
}

async function rmOne(
  accessor: DatabricksVolumeAccessor,
  path: PathSpec,
  opts: RmOpts,
  index: CommandOpts['index'],
): Promise<void> {
  let isDir = false
  try {
    const st = await dbxStat(accessor, path, index ?? undefined)
    isDir = st.type === FileType.DIRECTORY
  } catch (err) {
    if (opts.force) return
    throw err
  }
  if (isDir) {
    if (opts.recursive) {
      await dbxRmR(accessor, path)
    } else if (opts.removeDir) {
      const children = await dbxReaddir(accessor, path, index ?? undefined)
      if (children.length > 0) {
        throw new Error(`directory not empty: ${path.original}`)
      }
      await dbxRmdir(accessor, path)
    } else {
      throw new Error(`${path.original}: is a directory (use recursive=True)`)
    }
  } else {
    await dbxUnlink(accessor, path)
  }
}

async function rmCommand(
  accessor: DatabricksVolumeAccessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  if (paths.length === 0) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('rm: missing operand\n') })]
  }
  const resolved = await resolveGlob(accessor, paths, opts.index ?? undefined)
  const recursive = opts.flags.r === true || opts.flags.R === true
  const force = opts.flags.f === true
  const removeDir = opts.flags.d === true
  const verbose = opts.flags.v === true
  const verboseParts: string[] = []
  const writes: Record<string, Uint8Array> = {}
  for (const p of resolved) {
    try {
      await rmOne(accessor, p, { recursive, force, removeDir }, opts.index)
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err)
      return [null, new IOResult({ exitCode: 1, stderr: ENC.encode(`${msg}\n`) })]
    }
    writes[p.stripPrefix] = new Uint8Array()
    if (verbose) verboseParts.push(`removed '${p.original}'`)
  }
  const output: ByteSource | null = verbose ? formatRecords(verboseParts) : null
  return [output, new IOResult({ writes })]
}

export const DATABRICKS_VOLUME_RM = command({
  name: 'rm',
  resource: ResourceName.DATABRICKS_VOLUME,
  spec: specOf('rm'),
  fn: rmCommand,
  write: true,
})
